std/os/time.zig

```diff
@@ -0,0 +1,262 @@
+    //  should ever fail for us with CLOCK_REALTIME
+    var ts: posix.timespec = undefined;
+    const err = posix.clock_gettime(posix.CLOCK_REALTIME, &ts);
+    debug.assert(err == 0);
+    const sec_ms = ts.tv_sec * ms_per_s;
+    const nsec_ms = @divFloor(ts.tv_nsec, ns_per_s / ms_per_s);
+    return u64(sec_ms) + u64(nsec_ms);
+}
+
+/// Divisions of a second
+pub const ns_per_s = 1000000000;
+pub const us_per_s = 1000000;
+pub const ms_per_s = 1000;
+pub const cs_per_s = 100;
+
+/// Common time divisions
+pub const s_per_min = 60;
+pub const s_per_hour = s_per_min * 60;
+pub const s_per_day = s_per_hour * 24;
+pub const s_per_week = s_per_day * 7;
+
+
+/// A monotonic high-performance timer.
+/// Timer.start() must be called to initialize the struct, which captures
+///   the counter frequency on windows and darwin, records the resolution,
+///   and gives the user an oportunity to check for the existnece of
+///   monotonic clocks without forcing them to check for error on each read.
+/// .resolution is in nanoseconds on all platforms but .start_time's meaning 
+///   depends on the OS. On Windows and Darwin it is a hardware counter 
+///   value that requires calculation to convert to a meaninful unit.
+pub const Timer = struct {
+    
+    //if we used resolution's value when performing the
+    //  performance counter calc on windows, it would be
+    //  less precise
+    frequency: switch(builtin.os) {
+        Os.windows => u64,
+        Os.macosx, Os.ios => darwin.mach_timebase_info_data,
+        else => void,
+    },
+    resolution: u64,
+    start_time: u64,
+    
+    //Initialize the timer structure.
+    //This gives us an oportunity to grab the counter frequency in windows.
+    //On Windows: QueryPerformanceCounter will succeed on anything > XP.
+    //On Posix: CLOCK_MONOTONIC will only fail if the monotonic counter is not 
+    //  supported, or if the timespec pointer is out of bounds, which should be 
+    //  impossible here barring cosmic rays or other such occurances of
+    //  incredibly bad luck.
+    //On Darwin: This cannot fail, as far as I am able to tell.
+    const TimerError = error{TimerUnsupported};
+    pub fn start() TimerError!Timer {
+        var self: Timer = undefined;
+        
+        switch(builtin.os) {
+            Os.windows => {
+                var freq: i64 = undefined;
+                var err = windows.QueryPerformanceFrequency(&freq);
+                if(err == 0) return error.TimerUnsupported;
+                self.frequency = u64(freq);
+                self.resolution = @divFloor(ns_per_s, self.frequency);
+                var start_time: i64 = undefined;
+                _ = windows.QueryPerformanceCounter(&start_time);
+                self.start_time = u64(start_time);
+            },
+            Os.linux => {
+                var ts: posix.timespec = undefined;
+                var result = posix.clock_getres(posix.CLOCK_MONOTONIC, &ts);
```

Comment:
I think we want CLOCK_MONOTONIC_RAW here instead. The _RAW variant is the same but avoids any time skewing due to ntp and/or adjtime. This makes it more suitable as a high-performance timer since we won't get any sudden jumps/delays.

Windows' `QueryPerformanceCounter` isn't time-adjusted either so we match this behavior across operating systems. Not 100% on MacOS.

Reply:
Note that this is only on kernels 2.6.28 and newer and is linux-specific so we would need to keep that in mind:
https://linux.die.net/man/2/clock_gettime

Reply:
I'm actually not 100% convinced about this change after doing some more reading. So happy to hear disagreements.

Reply:
I had the same thought and did some reading on the subject. I went this way for two reasons:

1) CLOCK_MONOTONIC_RAW is linux specific from what I can tell. Easy enough to switch our way through that, but:
2) Pretty much everyone I've seen having this discussion has ultimately decided on using regular MONOTONIC. The reasoning is that even though the interval of CLOCK_MONOTONIC is not stable, it will average out to being more correct in relation to the real passage of time. 

Both are guaranteed to only increase (bugs aside) and neither should "jump" ahead, my understanding is that the skews applied at this level are small.

That said, I'm not entirely sure RAW isn't a better choice just because it matches the QPC behavior. My guess is that the Darwin behavior is also not subject to skew based on how it works.

Reply:
I think `CLOCK_MONOTONIC_RAW` is best for linux.
 * it's OK to depend on OS-specific things. we'll abstract darwin, linux, and windows (and others when supported) into a platform-agnostic API.
 * timer is trying to measure an absolute duration, rather than measure the difference between 2 clock times.

Reply:
Hmm, I see. 

This looks like a nice writeup: https://github.com/ros2/rcutils/issues/43

In this case I'm happy with `CLOCK_MONOTONIC`.

Reply:
Thanks for the comments @jorangreef. Based on this information I think we may want to consider an iteration on the std lib time API to take into consideration hardware suspension and perhaps force the user to make a choice.